### PR TITLE
Add support for cracking BitShares wallets

### PR DESCRIPTION
**Pull request by kholia**
Notes,

* I will find some time to tune OpenMP scaling tomorrow.

* I will also looking into cracking the backup `.bin` files soon.

### CODE REVIEW

File: src/bitshares_fmt_plug.c [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+john_register_one(&fmt_bitshares);
+#else
+
+#include <string.h>
+
+#ifdef _OPENMP
+#include <omp.h>
+#ifndef OMP_SCALE
+#define OMP_SCALE               8  // XXX
+#endif
```
Comment (magnumripper) [inline src/bitshares_fmt_plug.c]: The OMP_SCALE defines are to be placed outside the `#ifdef _OPENMP`

File: src/bitshares_fmt_plug.c [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+#include "options.h"
+#include "sha2.h"
+#include "aes.h"
+#include "memdbg.h"
+
+#define FORMAT_LABEL            "bitshares"
+#define FORMAT_NAME             "BitShares Wallet"
+#define FORMAT_TAG              "$BitShares$"
+#define TAG_LENGTH              (sizeof(FORMAT_TAG) - 1)
+#define ALGORITHM_NAME          "SHA-512 32/" ARCH_BITS_STR
```
Comment (magnumripper) [inline src/bitshares_fmt_plug.c]: This should be `64/`

File: src/bitshares_fmt_plug.c [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+static struct custom_salt {
+	uint32_t ctlen;
+	unsigned char ct[MAX_CIPHERTEXT_LENGTH];
+} *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+#ifdef _OPENMP
+	omp_autotune(self, OMP_SCALE);
+#endif
```
Comment (magnumripper) [inline src/bitshares_fmt_plug.c]: This should not be hidden in `#ifdef`

File: src/bitshares_fmt_plug.c [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
```
Comment (magnumripper) [inline src/bitshares_fmt_plug.c]: Should be `int index;`

How did you pick the template format? Next time I suggest you use one that has most things right...

File: src/bitshares_fmt_plug.c [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+}
+
+static int cmp_one(void *binary, int index)
+{
+	return cracked[index];
+}
+
+static int cmp_exact(char *source, int index)
+{
+	return cracked[index];
```
Comment (magnumripper) [inline src/bitshares_fmt_plug.c]: Should be `return 1;`

**Comment by kholia** [comment]
I used instructions from https://github.com/magnumripper/JohnTheRipper/issues/3091 to tune this format.

**Comment by kholia** [comment]
> I will also looking into cracking the backup .bin files soon.

This involves asymmetric crypto, and will be handled later in a new PR.

**Pull request merged**